### PR TITLE
Refactored building of TaskLifecycleListener

### PR DESCRIPTION
**Pull request by mminella**
When using Spring Boot's datasource initialization features, there was
the possibility that the datasource used by the TaskLifecycleListener
was not ready by the time it was needed for regular injection.  This
commit addresses that by obtaining the datasource at the last possible
moment.

Resolves spring-cloud/spring-cloud-task#83

### CODE REVIEW

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java [hunk 1]
```diff
@@ -45,64 +41,70 @@
...
-		else {
-			JdbcTaskRepositoryFactoryBean jdbcTaskRepositoryFactoryBean =
-					new JdbcTaskRepositoryFactoryBean(dataSource);
-			taskRepository = jdbcTaskRepositoryFactoryBean.getObject();
-			JdbcTaskExplorerFactoryBean jdbcTaskExplorerFactoryBean =
-					new JdbcTaskExplorerFactoryBean(dataSource);
-			taskExplorer = jdbcTaskExplorerFactoryBean.getObject();
-			transactionManager = new DataSourceTransactionManager(dataSource);
-		}
+	private boolean isDataSoureAvailable() {
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: isDataSourceAvailable is misspelled.

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java [hunk 2]
```diff
@@ -45,64 +41,70 @@
...
+	public DefaultTaskConfigurer(ConfigurableApplicationContext context) {
+		this.context = context;
 	}
 
 	@Override
 	public TaskRepository getTaskRepository() {
-		return taskRepository;
+		if(this.taskRepository == null) {
+			if(this.taskExecutionDaoFactoryBean == null) {
+				this.taskExecutionDaoFactoryBean = new TaskExecutionDaoFactoryBean(this.context);
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: Why not create the taskExecutionDAoFactoryBean in the ctor?

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java (hunk 1 above)
Comment (mminella) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: No it's not...`isDataSoureAvailable` is ;)

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java [hunk 3]
```diff
@@ -45,64 +41,70 @@
...
+			}
+
+			this.taskRepository = new SimpleTaskRepository(this.taskExecutionDaoFactoryBean);
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: SimpleTaskRepository can be initialized in the ctor as well

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java [hunk 4]
```diff
@@ -45,64 +41,70 @@
...
+		}
+
+		return this.taskRepository;
 	}
 
 	@Override
 	public TaskExplorer getTaskExplorer() {
+		if(this.taskExplorer == null) {
+			if(this.taskExecutionDaoFactoryBean == null) {
+				this.taskExecutionDaoFactoryBean = new TaskExecutionDaoFactoryBean(this.context);
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: Why not create the taskExecutionDAoFactoryBean in the ctor? *same comment as above*

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/DefaultTaskConfigurer.java]: Doh! foiled again :-)

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/SimpleTaskConfiguration.java [hunk 5]
```diff
@@ -92,16 +91,15 @@ public TaskNameResolver taskNameResolver() {
 	public TaskRepositoryInitializer taskRepositoryInitializer() {
 		TaskRepositoryInitializer taskRepositoryInitializer = new TaskRepositoryInitializer();
 
-		if(this.dataSources != null && this.dataSources.size() == 1) {
-			taskRepositoryInitializer.setDataSource(this.dataSources.iterator().next());
+		if(this.context.getBeanNamesForType(DataSource.class).length == 1) {
+			taskRepositoryInitializer.setDataSource(context.getBean(DataSource.class));
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/SimpleTaskConfiguration.java]: Should we throw a IllegalStateException if we find more than 1 DataSource?

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/support/TaskExecutionDaoFactoryBean.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		DataFieldMaxValueIncrementerFactory incrementerFactory = new DefaultDataFieldMaxValueIncrementerFactory(dataSource);
+		this.dao = new JdbcTaskExecutionDao(dataSource);
+		String databaseType;
+		try {
+			databaseType = org.springframework.batch.support.DatabaseType.fromMetaData(dataSource).name();
+		}
+		catch (MetaDataAccessException e) {
+			throw new IllegalStateException(e);
+		}
+		((JdbcTaskExecutionDao) this.dao).setTaskIncrementer(incrementerFactory.getIncrementer(databaseType, this.tablePrefix + "SEQ"));
```
Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/support/TaskExecutionDaoFactoryBean.java]: nitpicky (I know) but these can be in the try block.

Comment (mminella) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/support/TaskExecutionDaoFactoryBean.java]: Is there a benefit to that (I usually try to limit what's in there to only what can cause the exception)?

Comment (cppwfs) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/support/TaskExecutionDaoFactoryBean.java]: You can disregard.  Just looked awkward to me, but its ok.

**Comment by cppwfs** [comment]
LGTM Merging

**Comment by cppwfs** [comment]
Rebased Squashed Merged

**Pull request closed without merging**